### ISSUE TITLE
Cannot play data supplied by STDIN

### ISSUE THREAD

**Issue by eomanis** [celluloid-player/celluloid#542]
Overview Description:

Unlike `mpv`, `celluloid` cannot play data supplied by standard input / STDIN, which makes it impossible to use `celluloid` as media player for e.g. [streamlink](https://github.com/streamlink/streamlink).

Steps to Reproduce:
1. Play any live twitch.tv stream, for example [speedgaming](https://twitch.tv/speedgaming), with `streamlink` using its command line user interface and specifying celluloid as media player: `streamlink -p "celluloid -" twitch.tv/speedgaming best`

Actual Results: The `celluloid` window opens and displays an error message: 

> Error
> Playback was terminated abnormally. Reason: loading failed.

Expected Results: The `celluloid` window opens and starts playing the stream, same as when using `mpv` with e.g. `streamlink -p "mpv -" twitch.tv/speedgaming best`

Version: Arch Linux [celluloid package](https://www.archlinux.org/packages/community/x86_64/celluloid/), version `0.18-1`, build date 2019-11-09
Arch Linux [streamlink package](https://www.archlinux.org/packages/community/any/streamlink/), version `1.3.1-1`, build date 2020-01-28

Additional Information:
GNOME Shell `3.36` on Wayland, Linux `5.6.6-arch1-1` x86_64

**Comment by gnome-mpv** [comment]
It should work if you use named pipe instead: ` streamlink twitch.tv/speedgaming best -n --player=celluloid`

**Comment by eomanis** [comment]
> It should work if you use named pipe instead: ` streamlink twitch.tv/speedgaming best -n --player=celluloid`

You are right, it does, so at least this particular use case is covered after all.

Still, would be nice if plain STDIN would be possible, too. I guess this issue can be seen as feature request.

**Comment by NessInOnett** [comment]
Is there a specific technical reason why gnome mpv/celluloid doesn't support stdin by default? I've been curious about this for a long time, as a frequent streamlink user who hates mpv's auto-scaling player controls.  

Not a big deal as I can just use -n from now on, just wondering.  Glad to know that works!

**Comment by gnome-mpv** [comment]
> Is there a specific technical reason why gnome mpv/celluloid doesn't support stdin by default?

It's a bit tricky to implement due to single-instancing. This means that it's possible that the process reading the data and the process that's actually playing it could be different processes. If that's the case, we need to not only _stream_ the data across processes, but also have some kind of signaling system so that the process reading the file knows when it can exit.

**Comment by eomanis** [comment]
I have made the experience that even with the `streamlink -n` switch you can only ever have a single stream running at any time with `celluloid`, which is at times a bummer when you want to have something akin to picture-in-picture with streams.

Using `celluloid --new-window` does not do the trick either.
All it does is spawn an *additional* window showing a black video area, regardless of whether there are already other `celluloid` instances running or not.

> It's a bit tricky to implement due to single-instancing. This means that it's possible that the process reading the data and the process that's actually playing it could be different processes. For regular files, we can just send the filename to the process handling playback. This won't work for stdin, so we need to not only _stream_ the data across processes, but also have some kind of signaling system so that the process reading the file knows when it can exit.

Are you sure this is not overkill?
It might perhaps be easier to add e.g. a command line option `--new-instance` or `--isolated-instance` that just disables any attempts to work together with existing instances. That should cover this use case nicely I would think.

**Comment by gnome-mpv** [comment]
>I have made the experience that even with the streamlink -n switch you can only ever have a single stream running at any time with celluloid, which is at times a bummer when you want to have something akin to picture-in-picture with streams.
>
>Using celluloid --new-window does not do the trick either.
All it does is spawn an additional window showing a black video area, regardless of whether there are already other celluloid instances running or not.

This happens because `celluloid --new-window` exits immediately if there's a running instance, and since streamlink isn't aware that the stream is going to be used by another process, it exits too. For now, you can work around this problem by adjusting the player command so that it blocks if it finds a celluloid process.

```
streamlink twitch.tv/speedgaming best -n -p=sh -a="-c 'celluloid --new-window {filename} && tail --pid=`pgrep celluloid` -f /dev/null'"
```

>It might perhaps be easier to add e.g. a command line option --new-instance or --isolated-instance that just disables any attempts to work together with existing instances. That should cover this use case nicely I would think.

There's an undocumented option called `--no-existing-session` that does something similar. I left it undocumented because it's only intended to be used for debugging, and there are a few things that behave in unusual ways when it's enabled.

**Comment by Henry-Hiles** [comment]
This is also useful for piping from yt-dlp, to stream using yt-dlp config, it would be nice if this could be implemented :)
